### PR TITLE
added jsdoc extractor

### PR DESCRIPTION
**Pull request by igorkorsukov**
Extract from like this 

```cpp
#include <QUrl>

#include "interactiveapi.h"

using namespace muse::api;

/** APIDOC namespace: interactive
 * User interaction - informational messages, error messages, questions and other dialogs.
 * @namespace
 */

InteractiveApi::InteractiveApi(IApiEngine* e)
    : ApiObject(e)
{
}

/** APIDOC method: info(title, text)
 * Show information message
 * @param {String} title Title
 * @param {String} text Message
 */

void InteractiveApi::info(const QString& title, const QString& text)
{
    interactive()->info(title.toStdString(), text.toStdString());
}

/** APIDOC method: openUrl(url)
 * Open URL in external browser
 * @param {String} url URL
 */
void InteractiveApi::openUrl(const QString& url)
{
    interactive()->openUrl(QUrl(url));
}

```

To like this  
```js
/**
* User interaction - informational messages, error messages, questions and other dialogs.
* @namespace
*/
const interactive = {
	/**
	* Show information message
	* @param {String} title Title
	* @param {String} text Message
	*/
	info(title, text) {},

	/**
	* Open URL in external browser
	* @param {String} url URL
	*/
	openUrl(url) {},

};
```

### CODE REVIEW

File: tools/jsdoc/jsdoc_extractor/src/main.cpp [hunk 1]
```diff
@@ -0,0 +1,218 @@
...
+            scan(files, itemPath, filter);
+        } else if (fs::is_regular_file(itemPath)) {
+            if (filter.exts.empty() || contains(filter.exts, itemPath.extension().string())) {
+                files.push_back(itemPath);
+            }
+        }
+    }
+}
+
+static std::string extracDoc(const fs::path& fname)
```
Comment (cbjeukendrup) [inline tools/jsdoc/jsdoc_extractor/src/main.cpp:43]:
```suggestion
static std::string extractDoc(const fs::path& fname)
```

File: tools/jsdoc/jsdoc_extractor/src/main.cpp [hunk 2]
```diff
@@ -0,0 +1,218 @@
...
+{
+    std::ifstream f;
+    f.open(fname);
+
+    struct State {
+        bool nspaceStared = false;
```
Comment (cbjeukendrup) [inline tools/jsdoc/jsdoc_extractor/src/main.cpp:49]:
```suggestion
        bool namespaceStarted = false;
```

File: tools/jsdoc/jsdoc_extractor/src/main.cpp [hunk 3]
```diff
@@ -0,0 +1,218 @@
...
+
+    if (files.empty()) {
+        std::cout << "not found files\n";
+        return 0;
+    }
+
+    std::filesystem::create_directories(out);
+
+    for (const fs::path& f: files) {
+        if (f.stem() != "interactiveapi") {
```
Comment (cbjeukendrup) [inline tools/jsdoc/jsdoc_extractor/src/main.cpp:206]: This is temporary, right?

**Review by cbjeukendrup — APPROVED** [review]

File: tools/jsdoc/jsdoc_extractor/.gitignore [hunk 4]
```diff
@@ -0,0 +1,74 @@
+# This file is used to ignore files which are generated
```
Comment (shoogle) [inline tools/jsdoc/jsdoc_extractor/.gitignore:1]: Could the generated files go in a subdirectory, and then we ignore that directory?

File: src/framework/global/api/interactiveapi.cpp [hunk 5]
```diff
@@ -25,16 +25,31 @@
...
 {
 }
 
+/** APIDOC method: info(title, text)
+ * Show information message
+ * @param {String} title Title
+ * @param {String} text Message
+ */
+
 void InteractiveApi::info(const QString& title, const QString& text)
```
Comment (shoogle) [inline src/framework/global/api/interactiveapi.cpp:44]: Could we parse some of this from the method signature, perhaps?
```C++
/** APIDOC method
 * Show really, really long
 * information message
 */
void InteractiveApi::info(
    const QString& title, /// @param Title
    const QString& text,  /// @param Message
)
{
     interactive()->info(title.toStdString(), text.toStdString());
}
```

File: tools/jsdoc/jsdoc_extractor/src/utils.h [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+
+    while (current != std::string::npos) {
+        out.push_back(str.substr(previous, current - previous));
+        previous = current + delimLen;
+        current = str.find(delim, previous);
+    }
+    out.push_back(str.substr(previous, current - previous));
+}
+
+static bool startsWith(const std::string& str, const std::string& start)
```
Comment (shoogle) [inline tools/jsdoc/jsdoc_extractor/src/utils.h:48]: It seems strange to implement all this in C++ when a higher language like Python would give you it for free.

Comment (cbjeukendrup) [reply tools/jsdoc/jsdoc_extractor/src/utils.h:48]: Actually, I agree with this. This seems more a use case for Python. (Unless we eventually want to go down the route of using LLVM to build an AST and operate on that, but that seems way overkill to me.)

File: src/framework/global/api/interactiveapi.cpp (hunk 5 above)
Comment (igorkorsukov) [reply src/framework/global/api/interactiveapi.cpp:44]: For this approach, we will need a type map like this https://github.com/musescore/MuseScore/blob/4ba6bf9d403a303988fa3163813e4434603b3c7a/src/framework/extensions/devtools/apidumpmodel.cpp#L38  
This map will need to be kept up to date.

I think it's better like this: 
```
/** APIDOC method
 * Show information message
 * @param {String} title Title
 * @param {String} text Message
 */
void InteractiveApi::info(const QString& title, const QString& text)
```
`info(title, text)` - This part may be constructed, the name should be taken from the method signature, the arguments can be taken either from the signature or from the description. Additional checks can be added if we encounter a mismatch between the description and the signature (but I'm relying more on the review here)

**Pull request closed without merging**